index.go

```diff
@@ -102,17 +113,143 @@ func (i *rowIndexIter) Next() (sql.Row, error) {
 
 func (i *rowIndexIter) Close() error { return i.index.Close() }
 
+type indexKey interface {
+	encode() ([]byte, error)
+	decode([]byte) error
+}
+
 type packOffsetIndexKey struct {
 	Repository string
 	Packfile   string
 	Offset     int64
 	Hash       string
 }
 
-func init() {
-	gob.Register(sql.Row{})
-	gob.Register(time.Time{})
-	gob.Register([]interface{}{})
+func (k *packOffsetIndexKey) decode(data []byte) error {
+	buf := bytes.NewBuffer(data)
+	var err error
+
+	if k.Repository, err = readString(buf); err != nil {
+		return err
+	}
+
+	if k.Packfile, err = readHash(buf); err != nil {
+		return err
+	}
+
+	ok, err := readBool(buf)
+	if err != nil {
+		return err
+	}
+
+	if ok {
+		if k.Offset, err = readInt64(buf); err != nil {
+			return err
+		}
+		k.Hash = ""
+	} else {
+		k.Offset = -1
+		if k.Hash, err = readHash(buf); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (k *packOffsetIndexKey) encode() ([]byte, error) {
+	var buf bytes.Buffer
+	writeString(&buf, k.Repository)
+	if err := writeHash(&buf, k.Packfile); err != nil {
+		return nil, err
+	}
+	writeBool(&buf, k.Offset >= 0)
+	if k.Offset >= 0 {
+		writeInt64(&buf, k.Offset)
+	} else {
+		if err := writeHash(&buf, k.Hash); err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+func readInt64(buf *bytes.Buffer) (int64, error) {
+	var bs = make([]byte, 8)
+	_, err := io.ReadFull(buf, bs)
+	if err != nil {
+		return 0, fmt.Errorf("can't read int64: %s", err)
+	}
+
+	v, n := binary.Varint(bs)
+	if n <= 0 {
+		return 0, fmt.Errorf("no bytes were read trying to read int64")
+	}
+
+	return v, nil
+}
+
+func readString(buf *bytes.Buffer) (string, error) {
+	size, err := readInt64(buf)
+	if err != nil {
+		return "", fmt.Errorf("can't read string size: %s", err)
+	}
+
+	var b = make([]byte, int(size))
+	if _, err = io.ReadFull(buf, b); err != nil {
+		return "", fmt.Errorf("can't read string of size %d: %s", size, err)
+	}
+
+	return string(b), nil
+}
+
+func readBool(buf *bytes.Buffer) (bool, error) {
+	b, err := buf.ReadByte()
+	if err != nil {
+		return false, fmt.Errorf("can't read bool: %s", err)
+	}
+
+	return b == 1, nil
+}
+
+func writeInt64(buf *bytes.Buffer, n int64) {
+	var bs = make([]byte, 8)
```

Comment:
Do we want to use a `Varint` or a full size `Int64`? From the documentation the size of a `Varint` can be 10 bytes:

https://golang.org/src/encoding/binary/varint.go

Reply:
You'e right. I will change it to how it was before, storing a `Uint64`

Reply:
Done